mmtk/src/api.rs

```diff
@@ -1,32 +1,37 @@
+use crate::OpenJDK;
+use crate::OpenJDK_Upcalls;
+use crate::SINGLETON;
+use crate::UPCALLS;
 use libc::{c_char, c_void};
-use std::ffi::{CStr, CString};
-use std::lazy::SyncLazy;
 use mmtk::memory_manager;
+use mmtk::plan::barriers::BarrierSelector;
 use mmtk::scheduler::GCWorker;
 use mmtk::util::alloc::allocators::AllocatorSelector;
+use mmtk::util::alloc::is_alloced_by_malloc;
 use mmtk::util::constants::LOG_BYTES_IN_PAGE;
+use mmtk::util::options::PlanSelector;
 use mmtk::util::{Address, ObjectReference, OpaquePointer};
 use mmtk::AllocationSemantics;
-use mmtk::MutatorContext;
 use mmtk::Mutator;
+use mmtk::MutatorContext;
 use mmtk::{Plan, MMTK};
-use mmtk::util::options::PlanSelector;
-use mmtk::plan::barriers::BarrierSelector;
-use crate::OpenJDK;
-use crate::OpenJDK_Upcalls;
-use crate::SINGLETON;
-use crate::UPCALLS;
+use std::ffi::{CStr, CString};
+use std::lazy::SyncLazy;
 
 // Supported barriers:
 static NO_BARRIER: SyncLazy<CString> = SyncLazy::new(|| CString::new("NoBarrier").unwrap());
 static OBJECT_BARRIER: SyncLazy<CString> = SyncLazy::new(|| CString::new("ObjectBarrier").unwrap());
 
+extern {
+    static mut mmtk_enable_allocation_fastpath: bool;
```

Comment:
This is unnecessary. We can keep the flag `MMTK_ENABLE_ALLOCATION_FASTPATH`. For malloc allocator, we dont need to implement a fastpath. We just need to insert a slowpath call.

https://github.com/mmtk/mmtk-openjdk/blob/75296ac4300d405e38825c41e105404d71dd8a38/openjdk/mmtkBarrierSetAssembler_x86.cpp#L45
If it is malloc allocator, just jump to slow case.

https://github.com/mmtk/mmtk-openjdk/blob/75296ac4300d405e38825c41e105404d71dd8a38/openjdk/mmtkBarrierSetC2.cpp#L125
Move this line slightly earlier. If the allocator is malloc allocator, we set `always_slow` to `true` (like this: https://github.com/mmtk/mmtk-openjdk/blob/75296ac4300d405e38825c41e105404d71dd8a38/openjdk/mmtkBarrierSetC2.cpp#L80)

Reply:
It was me suggesting Paige to turn this constant into a global variable, and enable/disable it at vm boot time, depending on the selected plan. I think this is a more generic solution compare to adding a branch in assembler and c2 IR, especially when we have multiple plans that can't support allocation fast path in the future.

Reply:
Whether we have fastpath or not depends on the allocator type, not the plan. For example, for MS, we may allocate immortal objects via bump pointer allocator. In this case, we can still use bump pointer fastpath.  (I understand that for the current MS PR in mmtk-core, all allocation is done through malloc, which seems incorrect)

I didn't suggest working around the cases where fastpath is not *yet* supported. What I suggested is to properly implement for the malloc allocator, which basically has no fastpath - all we need is a call to the slowpath. Please correct me if I am wrong.